feat: Add head size 72 support for QKV Preprocessing kernel

qixiang-99 opened the pull request:
Head size 72 fmha kernels were added previously but only cross attention cases are tested. So we can reuse the added fmha kernels for head size 72 attention by removing the cross attention condition in checking.

However, with above changes, we also need to add head size 72 QKV Preprocessing kernel (v2) to support features like RoPE etc in the furture.

What this PR does: 
1. remove the cross attention condition for head size 72 fmha path
2. add head size 72 QKV Preprocessing kernel

qixiang-99 committed:
refactor: Fix headsize 72 attention error for TRTLLM attn backend in PyTorch workflow

- Remove the head size pre-check logic in AttentionOp because head size 72 can be supported with fmha kernels.
- Added support for head size 72 in unfused attention kernels(QKVPreprocessing).
- Enhanced unit tests by introducing a scenario generation function for better test coverage of attention configurations(include head size 72).

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #2964](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2964/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2964](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2964/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2078](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2078/) completed with status: 'FAILURE'

qixiang-99 committed:
update: Waive head_dim=72 test cases and enhance test representation

- Added a waiver for head_dim=72 cases on post sm100 in the test suite to address known issues.
- Introduced a custom __repr__ method in the Scenario class for pytest substring match.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --stage-list "B200_PCIe-PyTorch-2"

qixiang-99 commented:
/bot run --stage-list "B200_PCIe-PyTorch-2"

tensorrt-cicd commented:
[PR_Github #3206](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3206/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3206](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3206/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2228 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2228/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --stage-list "B200_PCIe-PyTorch-2"

tensorrt-cicd commented:
[PR_Github #3220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3220/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3220/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2238 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2238/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #3316](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3316/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --stage-list "B200_PCIe-PyTorch-2"

tensorrt-cicd commented:
[PR_Github #3326](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3326/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3326](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3326/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2320 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2320/) completed with status: 'SUCCESS'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #3330](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3330/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3330](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3330/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2324](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2324/) completed with status: 'SUCCESS'

symphonylyh commented:
is my understanding correctly that you're using the unfused attn kernel for the head size 72, and later on will switch to fused mha / trtllm-gen when they added 72 support?
